### ISSUE TITLE
Improve parser suggestion/error when using a builtin as a label

### ISSUE THREAD

**Issue by Profpatsch** [dhall-lang/dhall-haskell#494]
```
⊢ ({ Text = "foo"}).Text

Error: Invalid input

(stdin):1:4:
  |
1 | ({ Text = "foo"}).Text
  |    ^
unexpected 'T'
expecting '=', '}', or whitespace



⊢ ({ `Text` = "foo"}).Text

Error: Invalid input

(stdin):1:21:
  |
1 | ({ `Text` = "foo"}).Text
  |                     ^
unexpected 'T'
expecting '/'



⊢ ({ `Text` = "foo"}).`Text`

"foo"



⊢ ({ Integer/show = "foo"}).`Text`

Error: Invalid input

(stdin):1:4:
  |
1 | ({ Integer/show = "foo"}).`Text`
  |    ^
unexpected 'I'
expecting '=', '}', or whitespace
```

All of these should probably say something like: 
“unexpected builtin symbol `Text`/`Integer/show`, maybe you want to use it as record field label, in this case put it inside backticks, like: ```({ `Text` = "hello" }).`Text` ```

**Comment by Profpatsch** [comment]
One more example:

```
⊢ let Text = "hello" in Text

Error: Invalid input

(stdin):1:9:
  |
1 | let Text = "hello" in Text
  |         ^
expecting label or whitespace



⊢ let `Text` = "hello" in Text

Text



⊢ let `Text` = "hello" in `Text`

"hello"
```

though in this case it might make sense to forbid names that are builtins? Or at least to discourage it, see the second input, which parses but is not what was intended of course.

**Comment by Profpatsch** [comment]
One more contrived example that’s possible but probably not a good idea:

```
⊢ \(`Text`: Type) -> \(text : Text) -> \(notText : `Text`) -> "now I’m confused"

  λ(`Text` : Type)
→ λ(text : Text)
→ λ(notText : `Text`)
→ "now I\u2019m confused"
```

So, allow field (and union?) labels that are named after symbols, but maybe forbid them in function binding and let clauses.

**Comment by Gabriel439** [comment]
@Profpatsch: I would also like to keep the grammar for labels uniform throughout the language for conceptual consistency.  In particular, record fields should share the same grammar with bound variables of a `let` expression since a record is conceptually analogous to multiple `let` bindings, and `let`-bound variables should share the same grammar as `λ`-bound variables since many `let` expressions have an equivalent `λ`-encoding.

I don't think that naming bound variables after reserved keywords is a dangerous language feature if they are escaped with backticks.  For example, you can craft equally confusing examples without using backticks:

```haskell
λ(a : Type) → λ(a : a) → λ(a : a@1) → a

λ(Text_ : Type) → λ(text : Text) → λ(notText : Text_) -> "now I’m confused"
```

I can definitely improve the error message for parsing a reserved keyword, though.  I don't have a lot of flexibility since this is done in the parser rather than the type-checker, but it's still doable

**Comment by Gabriel439** [comment]
Actually, after taking a stab at this, I'm not sure that I can improve the parse error message.  Everything I tried did nothing to improve the error message :\

**Comment by f-f** [comment]
I think we can close this now that https://github.com/dhall-lang/dhall-lang/pull/437 is in?

**Comment by Gabriel439** [comment]
@f-f: It still needs to be implemented in the Haskell implementation (which I plan to do soon), to resolve this issue, since the issue is a complaint specifically about the Haskell implementation

**Comment by Gabriel439** [comment]
Closing, as this is fixed now (first available in version 1.22.0)
